pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -59,34 +65,62 @@ func NewScheduler(kubeClient client.Client) *Scheduler {
 func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) ([]*Node, error) {
 	defer metrics.Measure(schedulingDuration.WithLabelValues(injection.GetNamespacedName(ctx).Name))()
 	constraints := provisioner.Spec.Constraints.DeepCopy()
+	start := time.Now()
+
+	sort.Slice(pods, byCPUAndMemoryDescending(pods))
+	sort.Slice(instanceTypes, func(i, j int) bool {
+		return instanceTypes[i].Price() < instanceTypes[j].Price()
+	})
+
 	// Inject temporarily adds specific NodeSelectors to pods, which are then
 	// used by scheduling logic. This isn't strictly necessary, but is a useful
 	// trick to avoid passing topology decisions through the scheduling code. It
 	// lets us treat TopologySpreadConstraints as just-in-time NodeSelectors.
 	if err := s.topology.Inject(ctx, constraints, pods); err != nil {
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
-	return s.schedule(constraints, instanceTypes, pods), nil
-}
 
-// schedule separates pods into a list of TheoreticalNodes that contain the pods. All pods in each theoretical node
-// contain isomorphic scheduling constraints and can be deployed together on the same node, or multiple similar nodes if
-// the pods exceed one node's capacity.
-func (s *Scheduler) schedule(constraints *v1alpha5.Constraints, instanceTypes []cloudprovider.InstanceType, pods []*v1.Pod) []*Node {
-	var nodes []*Node
+	nodes, err := NewNodeSet(ctx, constraints, s.kubeClient)
```

Comment:
Consider naming this nodeSet so it reads `nodeSet.nodes` later in L90.

Reply:
sgtm